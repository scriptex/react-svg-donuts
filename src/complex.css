.donut-complex text {
	font-family: sans-serif;
	font-size: 0.75rem;
	opacity: 0;
	transition: opacity 0.35s ease-in-out 1s;
}

.donut-complex circle {
	transition: stroke-dashoffset 1s ease-in-out;
}

.donut-complex:not(.donut-complex--loaded) circle {
	stroke-dashoffset: 0;
}

.donut-complex--loaded text {
	opacity: 1;
}
