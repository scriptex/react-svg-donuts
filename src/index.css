@keyframes donut-fill {
	to {
		stroke-dasharray: 0 100;
	}
}

@keyframes text-slideup {
	from {
		opacity: 0;
		transform: translateY(0%);
	}
	to {
		opacity: 1;
		transform: translateY(-50%);
	}
}

.donut {
	width: 12.5rem;
	height: 12.5rem;
	position: relative;
	margin: auto;
}

.donut__frame,
.donut__circle {
	stroke-width: 0.125rem;
	fill: none;
}

.donut__frame {
	stroke: #ccc;
}

.donut__circle {
	stroke: #000;
	stroke-linecap: round;
}

.donut__canvas {
	width: 100%;
	height: 100%;
	display: block;
	position: relative;
	z-index: 2;
}

.donut__circle {
	transform: rotate(-90deg);
	transform-origin: 50% 50%;
	animation: donut-fill 2s reverse;
}

.donut__circle[stroke-dasharray='0 100'] {
	stroke: none;
}

.donut__text {
	font-family: monospace;
	font-size: 3rem;
	letter-spacing: -0.15rem;
	text-align: center;
	position: absolute;
	top: 50%;
	right: 0;
	left: 0;
	z-index: 3;
	transform: translateY(-50%);
	animation: text-slideup 2s linear;
}

.donut.is--negative .donut__circle {
	transform: rotate(-90deg) scaleY(-1);
}
