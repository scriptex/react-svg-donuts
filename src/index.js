import React from 'react';
import PropTypes from 'prop-types';

import './index.css';

const SIZE = 34;
const PREFIX = 'donut';

const halfSize = SIZE / 2;
const circleProps = {
	cx: halfSize,
	cy: halfSize,
	r: halfSize - 1
};

const getClassName = (p, c) => `${p}${c}`;
const renderProgress = progress => <strong>{progress}</strong>;

const Donut = ({ progress = 0, onRender = renderProgress, prefix = PREFIX }) => (
	<div className={getClassName(prefix, progress < 0 ? ' is--negative' : '')}>
		<svg
			className={getClassName(prefix, '__canvas')}
			width={SIZE}
			height={SIZE}
			viewBox={`0 0 ${SIZE} ${SIZE}`}
			xmlns="http://www.w3.org/2000/svg"
		>
			<circle className={getClassName(prefix, '__frame')} {...circleProps} />

			<circle
				className={getClassName(prefix, '__circle')}
				strokeDasharray={`${Math.abs(progress)} 100`}
				{...circleProps}
			/>
		</svg>

		{typeof onRender === 'function' && <div className={getClassName(prefix, '__text')}>{onRender(progress)}</div>}
	</div>
);

Donut.propTypes = {
	progress: PropTypes.number,
	onRender: PropTypes.func,
	prefix: PropTypes.string
};

Donut.defaultProps = {
	progress: 0,
	onRender: renderProgress,
	prefix: PREFIX
};

export { Donut };
export default Donut;
