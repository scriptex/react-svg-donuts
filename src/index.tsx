import * as React from 'react';

export interface DonutProps {
	prefix?: string;
	progress?: number;
	onRender?: (value: number) => React.ReactNode;
}

const SIZE = 34;
const PREFIX = 'donut';

const halfSize = SIZE / 2;
const circleProps = {
	cx: halfSize,
	cy: halfSize,
	r: halfSize - 1
};

const getClassName = (p: string, c: string): string => `${p}${c}`;
const renderProgress = (progress: number): React.ReactNode => <strong>{progress}</strong>;

export const Donut = ({ progress = 0, onRender = renderProgress, prefix = PREFIX }: DonutProps) => (
	<div className={getClassName(prefix, progress < 0 ? ' is--negative' : '')}>
		<svg
			className={getClassName(prefix, '__canvas')}
			width={SIZE}
			height={SIZE}
			viewBox={`0 0 ${SIZE} ${SIZE}`}
			xmlns="http://www.w3.org/2000/svg"
		>
			<circle className={getClassName(prefix, '__frame')} {...circleProps} />

			<circle
				className={getClassName(prefix, '__circle')}
				strokeDasharray={`${Math.abs(progress)} 100`}
				{...circleProps}
			/>
		</svg>

		{typeof onRender === 'function' && <div className={getClassName(prefix, '__text')}>{onRender(progress)}</div>}
	</div>
);

export type ComplexDonutTextCoords = {
	x: number;
	y: number;
};

export type ComplexDonutPart = {
	color: string;
	value: number;
};

export type ComplexDonutSegment = ComplexDonutPart & {
	rotate: string;
	percent: number;
	textCoords: ComplexDonutTextCoords;
};

export interface ComplexDonutProps {
	size?: number;
	parts?: ComplexDonutPart[];
	radius?: number;
	className?: string;
	thickness?: number;
	textProps?: React.SVGProps<SVGTextElement>;
	startAngle?: number;
	circleProps?: React.SVGProps<SVGCircleElement>;
}

let rotateAngle = 0;

const getTotal = (values: ComplexDonutPart[]): number => values.reduce((acc, { value }) => acc + value, 0);
const getPercent = (value: number, total: number): number => value / total;
const sortValues = (values: ComplexDonutPart[]): ComplexDonutPart[] => values.sort((a, b) => b.value - a.value);
const getCircumference = (radius: number): number => 2 * Math.PI * radius;
const convertDegreesToRadians = (angle: number): number => angle * (Math.PI / 180);

export const ComplexDonut: React.FC<Readonly<ComplexDonutProps>> = ({
	size = 160,
	parts = [],
	radius = 60,
	className = '',
	thickness = 30,
	textProps = {},
	startAngle = -90,
	circleProps = {}
}: ComplexDonutProps) => {
	const [segments, setSegments] = React.useState<ComplexDonutSegment[]>([]);
	const [isLoaded, setIsLoaded] = React.useState(false);

	const total = React.useMemo(() => getTotal(parts), [parts]);
	const halfSize = React.useMemo(() => size / 2, [size]);
	const circumference = React.useMemo(() => getCircumference(radius), [radius]);

	const getTextCoordinates = React.useCallback(
		(value: number, angleOffset: number) => {
			const total = getTotal(parts);
			const angle = (getPercent(value, total) * 360) / 2 + angleOffset;
			const radians = convertDegreesToRadians(angle);

			return {
				x: radius * Math.cos(radians) + size / 2,
				y: radius * Math.sin(radians) + size / 2
			};
		},
		[size, parts, radius]
	);

	const getTransforms = React.useCallback(() => {
		const rotations: number[] = [];
		const textCoords: Array<Record<'x' | 'y', number>> = [];

		const total = getTotal(parts);

		rotateAngle = startAngle;

		sortValues(parts).forEach(({ value }) => {
			const data = rotateAngle;
			const percent = getPercent(value, total);
			const { x, y } = getTextCoordinates(value, rotateAngle);

			rotations.push(data);
			textCoords.push({ x, y });

			const result = rotations[rotations.length - 1] || startAngle;

			rotateAngle = percent * 360 + result;
		});

		return { rotations, textCoords };
	}, [parts, startAngle, getTextCoordinates]);

	const getStrokeDashOffset = React.useCallback(
		(value: number, circumference: number) => circumference - getPercent(value, total) * circumference,
		[total]
	);

	React.useEffect(() => {
		const { rotations, textCoords } = getTransforms();

		setSegments(
			sortValues(parts).map(({ value, color }, i) => ({
				value,
				color,
				percent: getPercent(value, total),
				rotate: `rotate(${rotations[i]}, ${size / 2}, ${size / 2})`,
				textCoords: textCoords[i]
			}))
		);

		const loadTimeout = setTimeout(() => {
			setIsLoaded(true);
		}, 100);

		return () => {
			clearTimeout(loadTimeout);
		};
	}, [size, parts, total, getTransforms]);

	return (
		<div className={`donut-complex${isLoaded ? ' donut-complex--loaded ' : ' '}${className}`}>
			<svg height={size} width={size} viewBox={`0 0 ${size} ${size}`}>
				{segments.map((segment: ComplexDonutSegment, i: number) => (
					<g key={i}>
						<circle
							{...circleProps}
							r={radius}
							cx={halfSize}
							cy={halfSize}
							transform={segment.rotate}
							stroke={segment.color}
							strokeWidth={thickness}
							strokeDasharray={circumference}
							strokeDashoffset={getStrokeDashOffset(segment.value, circumference)}
						/>
						<text
							{...textProps}
							x={segment.textCoords.x}
							y={segment.textCoords.y}
							dy="3px"
							textAnchor="middle"
						>
							{`${Math.round(segment.percent * 100)}%`}
						</text>
					</g>
				))}
			</svg>
		</div>
	);
};
