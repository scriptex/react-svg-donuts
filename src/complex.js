import React from 'react';
import PropTypes from 'prop-types';

let rotateAngle = 0;

import './complex.css';

const getTotal = values => values.reduce((acc, { value }) => acc + value, 0);
const getPercent = (value, total) => value / total;
const sortValues = values => values.sort((a, b) => b.value - a.value);
const getCircumference = radius => 2 * Math.PI * radius;
const convertDegreesToRadians = angle => angle * (Math.PI / 180);

const ComplexDonut = props => {
	let loadTimeout;

	const total = getTotal(props.segments);

	const getTextCoordinates = (value, angleOffset) => {
		const { size, radius, segments } = props;
		const total = getTotal(segments);
		const angle = (getPercent(value, total) * 360) / 2 + angleOffset;
		const radians = convertDegreesToRadians(angle);

		return {
			x: radius * Math.cos(radians) + size / 2,
			y: radius * Math.sin(radians) + size / 2
		};
	};

	const getTransforms = () => {
		const rotations = [];
		const textCoords = [];
		const { startAngle, segments } = props;
		const total = getTotal(segments);

		rotateAngle = startAngle;

		sortValues(segments).forEach(({ value }) => {
			const data = rotateAngle;
			const percent = getPercent(value, total);
			const { x, y } = getTextCoordinates(value, rotateAngle);

			rotations.push(data);
			textCoords.push({ x, y });

			const result = rotations[rotations.length - 1] || startAngle;

			rotateAngle = percent * 360 + result;
		});

		return { rotations, textCoords };
	};

	const getStrokeDashOffset = (value, circumference) => {
		const diff = getPercent(value, total) * circumference;
		return circumference - diff;
	};

	const [segments, setSegments] = React.useState([]);
	const [isLoaded, setIsLoaded] = React.useState(false);

	React.useEffect(() => {
		const { segments, size } = props;
		const { rotations, textCoords } = getTransforms();

		setSegments(
			sortValues(segments).map(({ value, color }, i) => ({
				value,
				color,
				percent: getPercent(value, total),
				rotate: `rotate(${rotations[i]}, ${size / 2}, ${size / 2})`,
				textCoords: textCoords[i]
			}))
		);

		loadTimeout = setTimeout(() => {
			setIsLoaded(true);
		}, 100);

		return () => {
			clearTimeout(loadTimeout);
		};
	}, []);

	const { size, radius, thickness, className, circleProps, textProps } = props;
	const halfSize = size / 2;
	const circumference = getCircumference(radius);

	return (
		<div className={`donut-complex${isLoaded ? ' donut-complex--loaded ' : ' '}${className}`}>
			<svg height={size} width={size} viewBox={`0 0 ${size} ${size}`}>
				{segments.map((segment, i) => (
					<g key={i}>
						<circle
							{...circleProps}
							r={radius}
							cx={halfSize}
							cy={halfSize}
							transform={segment.rotate}
							stroke={segment.color}
							strokeWidth={thickness}
							strokeDasharray={circumference}
							strokeDashoffset={getStrokeDashOffset(segment.value, circumference)}
						/>
						<text
							{...textProps}
							x={segment.textCoords.x}
							y={segment.textCoords.y}
							dy="3px"
							textAnchor="middle"
						>
							{`${Math.round(segment.percent * 100)}%`}
						</text>
					</g>
				))}
			</svg>
		</div>
	);
};

ComplexDonut.propTypes = {
	size: PropTypes.number.isRequired,
	radius: PropTypes.number.isRequired,
	segments: PropTypes.arrayOf(
		PropTypes.shape({
			color: PropTypes.string,
			value: PropTypes.number
		})
	).isRequired,
	thickness: PropTypes.number.isRequired,
	startAngle: PropTypes.number,
	className: PropTypes.string,
	circleProps: PropTypes.object,
	textProps: PropTypes.object
};

ComplexDonut.defaultProps = {
	size: 160,
	radius: 60,
	segments: [],
	thickness: 30,
	startAngle: -90,
	className: '',
	circleProps: {},
	textProps: {}
};

export { ComplexDonut };
export default ComplexDonut;
