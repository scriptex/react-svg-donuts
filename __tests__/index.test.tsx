/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import { render, waitFor } from '@testing-library/react';

import { Donut, ComplexDonut } from '../src';

describe('Donuts', () => {
	it('should render simple donut with default props', () => {
		const { asFragment } = render(<Donut />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render a simple donut with custom props', () => {
		const { asFragment } = render(<Donut progress={50} onRender={progress => <strong>{progress}%</strong>} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render a complex donut', async () => {
		const { asFragment } = await waitFor(() =>
			render(
				<ComplexDonut
					size={200}
					radius={80}
					parts={[
						{
							color: '#FF8A80',
							value: 230
						},
						{
							color: '#FF80AB',
							value: 308
						},
						{
							color: '#B9F6CA',
							value: 520
						},
						{
							color: '#B388FF',
							value: 130
						},
						{
							color: '#8C9EFF',
							value: 200
						}
					]}
					thickness={40}
					startAngle={-90}
					textProps={{}}
					circleProps={{}}
				/>
			)
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
