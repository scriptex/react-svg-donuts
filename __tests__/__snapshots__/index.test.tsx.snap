// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Donuts should render a complex donut 1`] = `
<DocumentFragment>
  <div
    class="donut-complex "
  >
    <svg
      height="200"
      viewBox="0 0 200 200"
      width="200"
    >
      <g>
        <circle
          cx="100"
          cy="100"
          r="80"
          stroke="#B9F6CA"
          stroke-dasharray="502.6548245743669"
          stroke-dashoffset="314.34033698166456"
          stroke-width="40"
          transform="rotate(-90, 100, 100)"
        />
        <text
          dy="3px"
          text-anchor="middle"
          x="173.8756687233046"
          y="69.30170085029867"
        >
          37%
        </text>
      </g>
      <g>
        <circle
          cx="100"
          cy="100"
          r="80"
          stroke="#FF80AB"
          stroke-dasharray="502.6548245743669"
          stroke-dashoffset="391.11470500022784"
          stroke-width="40"
          transform="rotate(44.87031700288185, 100, 100)"
        />
        <text
          dy="3px"
          text-anchor="middle"
          x="107.23297432334213"
          y="179.6723545681805"
        >
          22%
        </text>
      </g>
      <g>
        <circle
          cx="100"
          cy="100"
          r="80"
          stroke="#FF8A80"
          stroke-dasharray="502.6548245743669"
          stroke-dashoffset="419.3618781391332"
          stroke-width="40"
          transform="rotate(124.75504322766571, 100, 100)"
        />
        <text
          dy="3px"
          text-anchor="middle"
          x="27.74388100642821"
          y="134.33734509228682"
        >
          17%
        </text>
      </g>
      <g>
        <circle
          cx="100"
          cy="100"
          r="80"
          stroke="#8C9EFF"
          stroke-dasharray="502.6548245743669"
          stroke-dashoffset="430.22617550025063"
          stroke-width="40"
          transform="rotate(184.4092219020173, 100, 100)"
        />
        <text
          dy="3px"
          text-anchor="middle"
          x="30.96065702741309"
          y="59.582564134849996"
        >
          14%
        </text>
      </g>
      <g>
        <circle
          cx="100"
          cy="100"
          r="80"
          stroke="#B388FF"
          stroke-dasharray="502.6548245743669"
          stroke-dashoffset="455.5762026761913"
          stroke-width="40"
          transform="rotate(236.28242074927954, 100, 100)"
        />
        <text
          dy="3px"
          text-anchor="middle"
          x="76.79888609122719"
          y="23.43820591579538"
        >
          9%
        </text>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Donuts should render a simple donut with custom props 1`] = `
<DocumentFragment>
  <div
    class="donut"
  >
    <svg
      class="donut__canvas"
      height="34"
      viewBox="0 0 34 34"
      width="34"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="donut__frame"
        cx="17"
        cy="17"
        r="16"
      />
      <circle
        class="donut__circle"
        cx="17"
        cy="17"
        r="16"
        stroke-dasharray="50 100"
      />
    </svg>
    <div
      class="donut__text"
    >
      <strong>
        50%
      </strong>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Donuts should render simple donut with default props 1`] = `
<DocumentFragment>
  <div
    class="donut"
  >
    <svg
      class="donut__canvas"
      height="34"
      viewBox="0 0 34 34"
      width="34"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="donut__frame"
        cx="17"
        cy="17"
        r="16"
      />
      <circle
        class="donut__circle"
        cx="17"
        cy="17"
        r="16"
        stroke-dasharray="0 100"
      />
    </svg>
    <div
      class="donut__text"
    >
      <strong>
        0
      </strong>
    </div>
  </div>
</DocumentFragment>
`;
